### ISSUE TITLE
Normalize batch shaping codebase-wide

### ISSUE THREAD

**Issue by ryanjulian** [rlworkgroup/garage#44]
Currently this is proxied by Policy.recurrent, but there are loss functions for non-recurrent policies which need fixed-length input trajectories/valid variables (i.e. any time you want to differentiate through the loss function).

**Comment by ryanjulian** [comment]
Proposal:
`(trajectory_index, time_step, data_dim_1, data_dim_2, data_dim_3)`

imagine I have 2D observations which are RGB images. i collect a bunch of them while training the robot, ending up with trajectories of images. the shape of a single image is `(channel, height, width)`. this would generate data of shape
`(trajectory_index, time_step, channel, height, width)`. This is the data shape which should be expected for all interfaces between garage modules. Internally, many modules might need to flatten this data, but it is usually most efficient to do this at the last possible instant.

flatten() should be used sparingly and most high-level primitives should expect dense input.

**Comment by ryanjulian** [comment]
@CatherineSue is this true in the TF tree?

I think @gautams3 was talking about this earlier.

**Comment by CatherineSue** [comment]
No. It is true for on-policy algorithms.

**Comment by ryanjulian** [comment]
Ah. Is this even relevant for off-policy algorithms? Is there a better default for replay buffers, for instance?

**Comment by gautams3** [comment]
Why is batch-shaping (or reshaping) irrelevant for off-policy algorithms?
@ryanjulian I like your proposal [here](https://github.com/rlworkgroup/garage/issues/44#issuecomment-408503797)
> Proposal:
> `(trajectory_index, time_step, data_dim_1, data_dim_2, data_dim_3)`

Also, `flatten()` should return flat trajectories (`(trajectory_index, time_step, flat_traj)`), instead of one long flat tensor (`(flat_data)`).

**Comment by ryanjulian** [comment]
That really depends on the context. What if you want to flatten everything but the batch dimension? That's usually what people mean when they say flatten in neural networks.

**Comment by CatherineSue** [comment]
For off-policy algorithms, if the algorithm uses a replay buffer, then the shape would be 
`(batch_size, data_dim1, data_dim2, data_dim3)`. There are exceptions, i.e. SAC can run with or without replay buffer.
